tool-plugins/nextjs-starter/src/hooks/useAppBridge.ts

```diff
@@ -0,0 +1,277 @@
+import {
+	AppBridgeSession,
+	BeginOAuthMessagePayload,
+	CreateBeginOAuthMessagePayload,
+	CreateValidateMessagePayload,
+	PluginType,
+	PostMessageAction,
+	ValidateMessagePayload,
+} from '@/types';
+import {
+	APP_BRIDGE_ORIGIN,
+	KEY_PARENT_HOST,
+	KEY_SLUG,
+	KEY_TOKEN,
+	KEY_VALIDATED_PAYLOAD,
+} from '@/utils/const';
+import { useState, useEffect } from 'react';
+
+const getPostMessageAction = (type: PluginType): PostMessageAction => {
+	switch (type) {
+		case 'space-plugin':
+			return 'app-changed';
+		case 'tool-plugin':
+			return 'tool-changed';
+		default:
+			throw new Error(`Invalid plugin type: ${type}`);
+	}
+};
+
+const getParentHost = () => {
+	const storedHost = sessionStorage.getItem(KEY_PARENT_HOST);
+	if (storedHost) {
+		return storedHost;
+	}
+	const params = new URLSearchParams(location.search);
+	const protocol = params.get('protocol');
+	const host = params.get('host');
+	if (!protocol || !host) {
+		throw new Error('Missing `protocol` or `host` in query params');
+	}
+	return `${protocol}//${host}`;
+};
+
+const getSlug = () => {
+	const storedSlug = sessionStorage.getItem(KEY_SLUG);
+	if (storedSlug) {
+		return storedSlug;
+	}
+	const params = new URLSearchParams(location.search);
+	return params.get('slug');
+};
+
+const postMessageToParent = (payload: unknown) => {
+	window.parent.postMessage(payload, getParentHost());
+};
+
+const useAppBridgeAuth = ({
+	type,
+	authenticated,
+}: {
+	type: PluginType;
+	authenticated: () => Promise<void>;
+}) => {
+	const [status, setStatus] = useState<
+		'init' | 'authenticating' | 'authenticated' | 'error'
+	>('init');
+	const [error, setError] = useState<unknown>();
+
+	const init = async () => {
+		const isInIframe = window.top !== window.self;
+
+		if (!isInIframe) {
+			setStatus('error');
+			setError('not-in-iframe');
+			return;
+		}
+
+		if (!isAuthenticated()) {
+			sendValidateMessageToParent();
+			return;
+		}
+
+		setStatus('authenticated');
+		setError(undefined);
+
+		await authenticated();
+	};
+
+	const isAuthenticated = () => {
+		try {
+			const payload: AppBridgeSession = JSON.parse(
+				sessionStorage.getItem(KEY_VALIDATED_PAYLOAD) || '',
+			);
+			return payload && new Date().getTime() / 1000 < payload.exp;
+		} catch (err) {
```

Comment:
nit: 
```suggestion
		} catch {
```
or
```suggestion
		} catch (_err) {
```

not important, but wdyt? 🤔